pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,963 @@
+	) -> Result<u128, DispatchError> {
+		let max_limit_from_spot_price = Self::get_max_limit_with_slippage(asset_in, asset_out, amount_out)?;
+		let max_limit = max(max_limit, &max_limit_from_spot_price);
+
+		let fee_amount_in_sold_asset = Self::get_transaction_fee(*asset_in)?;
+		let amount_to_sell_plus_fee = max_limit
+			.checked_add(&fee_amount_in_sold_asset)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		Ok(amount_to_sell_plus_fee)
+	}
+
+	fn get_storage_bond_in_sold_currency(order: &Order<<T as Config>::Asset>) -> Result<Balance, DispatchError> {
+		let sold_currency = Self::sold_currency(order);
+		let storage_bond_in_native_currency = T::StorageBondInNativeCurrency::get();
+
+		let storage_bond_in_user_currency =
+			Self::convert_to_currency_if_asset_is_not_native(sold_currency, storage_bond_in_native_currency)?;
+
+		Ok(storage_bond_in_user_currency)
+	}
+
+	fn get_on_initialize_weight() -> u64 {
+		crate::weights::HydraWeight::<T>::on_initialize().ref_time()
+	}
+
+	fn get_execute_schedule_weight() -> u64 {
+		crate::weights::HydraWeight::<T>::execute_schedule().ref_time()
+	}
+
+	fn take_transaction_fee_from_user(owner: &T::AccountId, order: &Order<<T as Config>::Asset>) -> DispatchResult {
+		let fee_currency = Self::sold_currency(order);
+
+		let fee_amount_in_sold_asset = Self::get_transaction_fee(fee_currency)?;
+
+		T::Currency::transfer(
+			fee_currency.into(),
+			owner,
+			&T::FeeReceiver::get(),
+			fee_amount_in_sold_asset.into(),
+		)?;
+
+		Ok(())
+	}
+
+	fn unreserve_all_named_reserved_sold_currency(schedule_id: ScheduleId, who: &T::AccountId) -> DispatchResult {
+		let schedule = Schedules::<T>::get(schedule_id).ok_or(Error::<T>::ScheduleNotExist)?;
+		let named_reserve_identitifer = reserve_identifier(schedule_id);
+		let sold_currency = Self::sold_currency(&schedule.order);
+		T::Currency::unreserve_all_named(&named_reserve_identitifer, sold_currency.into(), who);
+
+		Ok(())
+	}
+
+	fn sold_currency(order: &Order<T::Asset>) -> <T as Config>::Asset {
+		let sold_currency = match order {
+			Order::Sell { asset_in, .. } => asset_in,
+			Order::Buy { asset_in, .. } => asset_in,
+		};
+		*sold_currency
+	}
+
+	fn weight_to_fee(weight: Weight) -> Balance {
+		// cap the weight to the maximum defined in runtime, otherwise it will be the
+		// `Bounded` maximum of its data type, which is not desired.
+		let capped_weight: Weight = weight.min(T::BlockWeights::get().max_block);
+		<T as pallet::Config>::WeightToFee::weight_to_fee(&capped_weight)
+	}
+
+	fn add_schedule_id_to_existing_ids_per_block(
+		next_schedule_id: ScheduleId,
+		blocknumber_for_schedule: <T as frame_system::Config>::BlockNumber,
+	) -> DispatchResult {
+		let schedule_ids =
+			ScheduleIdsPerBlock::<T>::get(blocknumber_for_schedule).ok_or(Error::<T>::NoScheduleIdsPlannedInBlock)?;
+		if schedule_ids.len() == T::MaxSchedulePerBlock::get() as usize {
+			let mut consequent_block = blocknumber_for_schedule;
+			consequent_block.saturating_inc();
```

Comment:
perhaps you should let it overflow and catch with an error? otherwise once it reaches MAX it will stop incrementing, and you will stay at the same block

Reply:
I was discussing this with Frido, and he said we will reach this roughly in 817 years given 6s blocktime u32 type. So I guess we are just fine with this?!

Reply:
but you are right, it is better be safe than sorry, I change to to checked_add